ui/org.eclipse.pde.core/src/org/eclipse/pde/internal/core/TargetPlatformPluginSourcePathLocator.java

```diff
@@ -0,0 +1,75 @@
+/*******************************************************************************
+ * Copyright (c) 2022 Christoph Läubrich and others.
+ *
+ * This program and the accompanying materials
+ * are made available under the terms of the Eclipse Public License 2.0
+ * which accompanies this distribution, and is available at
+ * https://www.eclipse.org/legal/epl-2.0/
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ *
+ * Contributors:
+ *     Christoph Läubrich - initial API and implementation
+ *******************************************************************************/
+package org.eclipse.pde.internal.core;
+
+import java.io.File;
+import java.net.URI;
+import org.eclipse.core.runtime.CoreException;
+import org.eclipse.core.runtime.IPath;
+import org.eclipse.core.runtime.Path;
+import org.eclipse.equinox.frameworkadmin.BundleInfo;
+import org.eclipse.pde.core.IPluginSourcePathLocator;
+import org.eclipse.pde.core.plugin.IPluginBase;
+import org.eclipse.pde.core.target.ITargetPlatformService;
+import org.eclipse.pde.core.target.TargetBundle;
+import org.eclipse.pde.internal.core.target.TargetPlatformService;
+
+/**
+ * Investigates the current active target if a matching source bundle can be
+ * found.
+ */
+public class TargetPlatformPluginSourcePathLocator implements IPluginSourcePathLocator {
+
+	@Override
+	public IPath locateSource(IPluginBase plugin) {
+		ITargetPlatformService platformService = TargetPlatformService.getDefault();
+		try {
+			TargetBundle[] bundles = platformService.getWorkspaceTargetDefinition().getBundles();
```

Comment:
Can't you build a mapping from plugin to source bundle once and re-use it until the TP is re-loaded? IIRC one can listen to this event.
Then this operation would be very cheap after the cache has been build.

But at the moment I'm not 100% certain what the key of such mapping would be (since IPluginBase is not a BundleInfo or TargetBundle), but probably the bundle file path or similar would be a good start.

Reply:
This is something one might maybe more implement in the Targetplatform itself because having such state in each component using it and especially with extensions that do not have a lifecycle is often cumbersome, for the moment I think it does not matter much and can be improved later, if one is certain work could even be parallelized here :-)

I even first did a Map-like approach here https://github.com/eclipse-pde/eclipse.pde/pull/409 but in the end it didn't worked out so well and was hard to understand (and didn't solved the caching issue anyways).

Reply:
For small targets it probably does not matter, but for large there could be a noticeable impact. But I'm neither sure about that nor do I have experience in this exact area and it just an guess and actually one would have to make some performance measurements. But if it isn't too complicated it should be addressed.

Couldn't we add a method `getBundles(String symbolicName)` that returns all bundles with that name and the TargetDefinition maintains the mapping from name -> TargetBundles internally?